### PR TITLE
Correct response codes and messages related to private factory flow

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
To support OAuth flow for private repository factories,  Che server must return correct responses to the clients. Most 
important use case is iniitating the OAuth procedure. 

 - User accepts factory from private repository;

 - Dashboard making request to resolve factory details using the reposiroty URL which is provided by user;
(POST /api/factory/resolver)

- Che server recognizes that URL points to private repository and looks for the credentials secret; If no secret present,
initiates the OAuth login procedure by returning the following response:

`POST https://<che-host>/api/factory/resolver`:

```
Response Status: 401

Response body:

 {
 "attributes": {
 "oauth_provider": "<provider>",
 "oauth_version": "<1.0|2.0>"
 "oauth_authentication_url": "https://<che-host>/api/oauth/<1.0>/authenticate?oauth_provider=<provider>&request_method=<GET|POST>&signature_method=<rsa|ssha>"
 },
 "errorCode": 401,
 "message": "SCM Authentication required"
}
```

Please note that parameters `token` and `redirect_after_login` are not returned and must be set by client on its discretion.

Other use cases responses:

OAuth not confitured:

```
Response Status:  500

Message: Provided location is unknown or misconfigured on the server side. Error message: <>
```

Cannot communicate with OAuth server:

```
Response Status:  500

Message: There is an error happened when communicate with SCM server. Error message: <>
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: core/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java [hunk 1]
```diff
@@ -1,38 +0,0 @@
...
-package org.eclipse.che.api.core;
-
-/**
- * Error codes that are used in exceptions. Defined error codes MUST BE in range <b>15000-32999</b>
- * inclusive.
- *
- * @author Igor Vinokur
- */
-public class ErrorCodes {
-  private ErrorCodes() {}
```
Comment (skabashnyuk) [inline core/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java:21]: what is the reason to move it to a different location?

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/NoopOAuthAuthenticator.java [hunk 2]
```diff
@@ -46,4 +46,9 @@ String computeAuthorizationHeader(String userId, String requestMethod, String re
     throw new RuntimeException(
         "The fallback noop authenticator cannot be used for authentication. Make sure OAuth is properly configured.");
   }
+
+  @Override
+  public String getLocalAuthenticateUrl() {
+    return null;
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/NoopOAuthAuthenticator.java]: RuntimeException ?

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java [hunk 3]
```diff
@@ -120,17 +121,38 @@ public URLFactoryBuilder(
...
+      } catch (OverrideParameterException e) {
+        throw new BadRequestException("Error processing override parameter(s): " + e.getMessage());
+      } catch (DevfileException e) {
+        throw toApiException(e, location);
       }
     }
     return Optional.empty();
   }
 
+  private ApiException toApiException(DevfileException devfileException, DevfileLocation location) {
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java:134]: tests?

**Review by skabashnyuk — COMMENTED** [review]
looks good in general

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/NoopOAuthAuthenticator.java (hunk 2 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/NoopOAuthAuthenticator.java]: \+

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java (hunk 3 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java:134]: Added

File: core/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java (hunk 1 above)
Comment (mshaposhnik) [reply core/che-core-api-core/src/main/java/org/eclipse/che/api/core/ErrorCodes.java:21]: It is removed as unused anymore

**Pull request closed without merging**